### PR TITLE
Add TypeScript subsection for atomFamily

### PR DESCRIPTION
**Pull request by dai-shi**
close https://github.com/pmndrs/jotai/issues/436

### CODE REVIEW

File: docs/jotai/api/utils.mdx [hunk 1]
```diff
@@ -222,6 +222,33 @@ import deepEqual from 'fast-deep-equal'
 const fooFamily = atomFamily((param) => atom(param), deepEqual)
 ```
 
+### TypeScript
+
+Types are defined like this:
```
Comment (HosseinAgha) [inline docs/jotai/api/utils.mdx:227]: Maybe change this line to:  
"You can explicitly declare the parameter, value, and atom's update function types using TypeScript generics"

File: docs/jotai/api/utils.mdx [hunk 2]
```diff
@@ -222,6 +222,33 @@ import deepEqual from 'fast-deep-equal'
...
+
+```ts
+atomFamily<Param, Value, Update>(initializeAtom: (param: Param) => WritableAtom<Value, Update>, areEqual?: (a: Param, b: Param) => boolean)
+atomFamily<Param, Value>(initializeAtom: (param: Param) => Atom<Value>, areEqual?: (a: Param, b: Param) => boolean)
+```
+
+It's generally recommended to infer types from initializeAtom.
```
Comment (HosseinAgha) [inline docs/jotai/api/utils.mdx:234]: "The atom family types will be inferred from initializeAtom."  

I think we can mention this part first and then talk about generics.

**Review by HosseinAgha — COMMENTED** [review]
Thank you for the PR.  
I hope this change makes using atomFamily more clear.

**Comment by dai-shi** [comment]
@HosseinAgha I made changes based on your comments. Not sure if they are aligned. Feel free to add suggestions (if you use GitHub "suggestions" feature, it's easier.)

File: docs/jotai/api/utils.mdx [hunk 3]
```diff
@@ -222,6 +222,35 @@ import deepEqual from 'fast-deep-equal'
...
+
+```ts
+import type { PrimitiveAtom } from 'jotai'
+
+const myFamily = atomFamily((id: number) => atom(id))
+// the type of atom(id) is PrimitiveAtom<number>
+// PrimitiveAtom<number> is WritableAtom<number, SetStateAction<number>>
+```
+
+If you want to explicitly declare it for a primitive atom, you need to use `SetStateAction`.
```
Comment (CodyJasonBennett) [inline docs/jotai/api/utils.mdx:246]: `it` is a bit ambiguous here. Would it help if we explicitly state what we're talking about?

Comment (dai-shi) [reply docs/jotai/api/utils.mdx:246]:
```suggestion
If you want to explicitly declare the atomFamily for a primitive atom, you need to use `SetStateAction`.
```
Does this help?

Comment (CodyJasonBennett) [reply docs/jotai/api/utils.mdx:246]: I believe so, yes.

**Review by CodyJasonBennett — APPROVED** [review]

File: docs/jotai/api/utils.mdx (hunk 3 above)
Comment (CodyJasonBennett) [reply docs/jotai/api/utils.mdx:246]: I believe so, yes.

File: docs/jotai/api/utils.mdx [hunk 4]
```diff
@@ -222,6 +222,35 @@ import deepEqual from 'fast-deep-equal'
...
+
+Here's a typical usage with a primitive atom.
+
+```ts
+import type { PrimitiveAtom } from 'jotai'
+
+const myFamily = atomFamily((id: number) => atom(id))
+// the type of atom(id) is PrimitiveAtom<number>
+// PrimitiveAtom<number> is WritableAtom<number, SetStateAction<number>>
+```
```
Comment (HosseinAgha) [inline docs/jotai/api/utils.mdx:244]:
```suggestion
The atom family types will be inferred from initializeAtom. Here's a typical usage with a primitive atom.

```ts
import type { PrimitiveAtom } from 'jotai'

/** 
 * here the atom(id) returns a PrimitiveAtom<number>
 * and PrimitiveAtom<number> is a WritableAtom<number, SetStateAction<number>>
 */ 
const myFamily = atomFamily((id: number) => atom(id))
```

You can explicitly declare the type of parameter, value, and atom's setState function using TypeScript generics.

```ts
atomFamily<Param, Value, Update>(initializeAtom: (param: Param) => WritableAtom<Value, Update>, areEqual?: (a: Param, b: Param) => boolean)
atomFamily<Param, Value>(initializeAtom: (param: Param) => Atom<Value>, areEqual?: (a: Param, b: Param) => boolean)
```
```

Comment (HosseinAgha) [reply docs/jotai/api/utils.mdx:244]: looks like Github has a bug when adding code inside suggestions. I had to artificially escape "```".

Comment (dai-shi) [reply docs/jotai/api/utils.mdx:244]: Thanks. OK, I'll commit it and fix it.

**Comment by dai-shi** [comment]
Thanks @HosseinAgha !
Let me merge this. Feel free to open a PR if you or someone has follow-up suggestions.

**Pull request merged**